be/src/io/async_flush_output_stream.cpp

```diff
@@ -0,0 +1,161 @@
+    {
+        while (!_slice_chunk_queue.empty() && _slice_chunk_queue.front()->is_full()) {
+            auto chunk = _slice_chunk_queue.front();
+            _slice_chunk_queue.pop_front();
+            _releasable_bytes.fetch_add(chunk->get_buffer_ptr()->capacity());
+
+            auto task = [this, chunk]() {
+                SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(_runtime_state->instance_mem_tracker());
+                CurrentThread::current().set_query_id(_runtime_state->query_id());
+                CurrentThread::current().set_fragment_instance_id(_runtime_state->fragment_instance_id());
+                auto buffer = chunk->get_buffer_ptr();
+                DeferOp op([this, chunk, capacity = buffer->capacity()] {
+                    _releasable_bytes.fetch_sub(capacity);
+                    delete chunk;
+                });
+                auto status = _file->append(Slice(buffer->data(), buffer->size()));
+                {
+                    std::scoped_lock lock(_mutex);
+                    _io_status.update(status);
+                    if (_task_queue.empty()) {
+                        _has_in_flight_io = false;
+                        return;
+                    }
+                    auto task = _task_queue.front();
+                    _task_queue.pop();
+                    bool ok = _io_executor->offer(task);
+                    DCHECK(ok);
+                }
+            };
+            to_enqueue_tasks.emplace_back(task);
+        }
+    }
+
+    if (!to_enqueue_tasks.empty()) {
+        enqueue_tasks_and_maybe_submit_task(std::move(to_enqueue_tasks));
+    }
+    return Status::OK();
+}
+
+Status AsyncFlushOutputStream::close() {
+    DCHECK(_slice_chunk_queue.empty() || (_slice_chunk_queue.size() == 1 && !_slice_chunk_queue.front()->is_full()))
+            << "empty or at most one not full buffer";
+
+    std::vector<Task> to_enqueue_tasks;
+    if (!_slice_chunk_queue.empty() && !_slice_chunk_queue.front()->is_empty()) {
+        auto chunk = _slice_chunk_queue.front();
+        _slice_chunk_queue.pop_front();
+        _releasable_bytes.fetch_add(chunk->get_buffer_ptr()->capacity());
+        auto task = [&, chunk]() {
+            SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(_runtime_state->instance_mem_tracker());
+            CurrentThread::current().set_query_id(_runtime_state->query_id());
+            CurrentThread::current().set_fragment_instance_id(_runtime_state->fragment_instance_id());
+            auto buffer = chunk->get_buffer_ptr();
+            DeferOp op([this, chunk, capacity = buffer->capacity()] {
+                _releasable_bytes.fetch_sub(capacity);
+                delete chunk;
+            });
+            auto status = _file->append(Slice(buffer->data(), buffer->size()));
+            {
+                std::scoped_lock lock(_mutex);
+                _io_status.update(status);
+                if (_task_queue.empty()) {
+                    _has_in_flight_io = false;
```

Comment:
when will the _task_queue is not empty after the last chunk flushed?

Reply:
there is a close task (T2) following the last flush task (T1).

The T1 may or may not be done before the submission of T2.

Reply:
got it